scipy/optimize/tests/test_optimize.py

```diff
@@ -590,6 +591,37 @@ def test_minimize_l_bfgs_maxls(self):
                                 options={'disp': False, 'maxls': 1})
         assert_(not sol.success)
 
+    def test_minimize_l_bfgs_b_maxfun_interruption(self):
+        # gh-6162
+        f = optimize.rosen
+        g = optimize.rosen_der
+        x0 = np.ones(7) * 1000
+
+        class _F(object):
+
+            def __init__(self):
+                self.values = []
+
+            def __call__(self, x):
+                value = f(x)
+                self.values.append(value)
+                return value
```

Comment:
You could define the `values` list outside and have a wrapper function to rosen that updates it. That would avoid this class.
```
values = []

def objfun(x):
    value = f(x)
    values.append(value)
    return value
```